### PR TITLE
Make all core scss files available from designsystem by using forward rule

### LINKED ISSUE #1939: [Housekeeping] @forward all Sass files instead of copying
- [x] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the housekeeping request

Copying Sass files (including their content) from `core` to `designsystem` as we currently do introduces a few problems. First and foremost a bug was introduced in relationsbanken due do this (configuring the same module two times). Fixed by #1936.

**Idea for how to handle scss files**

We could create a copy of each file in core with the same name and file position in the designsystem package. The content of the copy should then be replaced with a `@forward <file name>` rule.

This would mean that the core scss are the source of truth and we do not have to change names two places if we have to rename them.

### PR DESCRIPTION
**Pull request by MadsBuchmann**
## Which issue does this PR close?

This PR closes #1939

## What is the new behavior?

A script that is executed when publishing has been created which forwards all scss files from `libs/core` to the designsystem distribution bundle. 

To test it out execute: `npm run publish`. 

The distribution has a `scss` folder at the following path with files forwarding the corresponding file from core scss: `dist/libs/designsystem/scss`. 

Example`cat dist/libs/designsystem/scss/print/components/_kirby-page.scss`: 
> @forward "~@kirbydesign/core/scss/print/components/kirby-page";

Also the `copyScssFiles` step in `publish.js` has been removed as it is unnecessary with the forwarded files.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] ~~Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".~~
- [ ] ~~Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).~~

### Review  
- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [ ] Request that the changes are code-reviewed 
- [ ] ~~Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)~~

When the pull request has been approved it will be automatically merged to master via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

### CODE REVIEW

File: copy-scss-files.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+function createForwardScssFile(sourceFilePath, rootDir, targetFilePath) {
+  const newFileName = targetFilePath + sourceFilePath.split(rootDir).pop();
+
+  const newFileFolder = path.dirname(newFileName);
+  if (!fs.existsSync(newFileFolder)) {
+    fs.mkdirSync(newFileFolder, { recursive: true });
+  }
+
+  const forwardLocation = sourceFilePath.replace('./libs', '@kirbydesign');
```
Comment (RasmusKjeldgaard) [inline copy-scss-files.js:32]:
```suggestion
  const forwardLocation = sourceFilePath.replace('./libs', '~@kirbydesign');
```
I believe there should be a `~` here so we target the styles in the node_modules package correctly.

If I add it to the generated files, it works!

**Review by jkaltoft — CHANGES_REQUESTED** [review]
Noice!

Usually when you use [URLs with Sass](https://sass-lang.com/documentation/at-rules/use#finding-the-module) you do not include the extension, e.g., instead of

```
@forward "@kirbydesign/core/scss/_utils.scss";
```

it should output

```
@forward "@kirbydesign/core/scss/utils";
```

Also, when a file is named `index.scss` you can omit it, e.g.

```
@forward "@kirbydesign/core/scss/print/components/_index.scss";
```

should be

```
@forward "@kirbydesign/core/scss/print/components";
```

Is it possible to output proper URLs when `@forward`-ing?

File: copy-scss-files.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  return ext[ext.length - 1];
+}
+
+function createForwardRule(filePath, packageAlias, libsFolder) {
+  return pipe(
+    (fileName) => (fileName[0] === '_' ? fileName.slice(1) : fileName),
+    (fileName) => (fileName === 'index.scss' ? '' : fileName),
+    (fileName) =>
+      `@forward "~${path.dirname(filePath).replace(libsFolder, packageAlias)}/${fileName}";`
+  )(path.basename(filePath));
```
Comment (MadsBuchmann) [inline copy-scss-files.js:25]: remove appended slash if fileName is empty. Remove extension from fileName.

File: scripts/forward-scss-files.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  const targetFileDirName = path.dirname(targetFilePath);
+  if (!fs.existsSync(targetFileDirName)) {
+    fs.mkdirSync(targetFileDirName, { recursive: true });
+  }
+
+  const forwardRule = createForwardRule(sourceFilePath, packageAlias, sharedRootDir);
+  fs.writeFileSync(targetFilePath, forwardRule);
+}
+
+module.exports.forwardScssFiles = (sourceRootDir, targetRootDir, packageAlias, sharedRootDir) => {
```
Comment (jkaltoft) [inline scripts/forward-scss-files.js:59]: Same as mentioned before (object literal). Also totally optional.

File: scripts/publish.js [hunk 4]
```diff
@@ -112,31 +118,27 @@ function copyReadme() {
...
-
-function copyScssUtilsFile() {
-  console.log('Copying SCSS utils file...');
-  const onlyScssFiles = (input) => ['', '.scss'].includes(path.extname(input));
-  return fs.copy(`${angularLibDir}/scss/_utils.scss`, `${distTarget}/scss/_utils.scss`, {
-    filter: onlyScssFiles,
+function forwardCoreScssFilesToDesignsystem() {
+  console.log('Forwarding core SCSS files to designsystem ...');
+  return new Promise((resolve) => {
+    forwardScssFiles(`${coreLibDir}/scss`, `${distTarget}/scss`, packageAlias, libsRootDir);
```
Comment (jkaltoft) [inline scripts/publish.js:124]: Optional.

Just a thought: Would it be more "future proof" to

* rename this function to `forwardCoreScssFiles()` and
* pass the destination as argument

or would that be a premature optimization?

**Review by jkaltoft — CHANGES_REQUESTED** [review]

**Review by jkaltoft — COMMENTED** [review]
It seems like something is a bit buggy.

I did this:

1. Create 2 new files in `libs/core/src/scss`
    - `awesome-stylesheet.scss`
    - `_awesome-partial.scss`
2. `npm run publish`

image

This resulted in the following output in `dist/libs/designsystem/scss/`:

image

**Review by jkaltoft — COMMENTED** [review]
Ok, this is weird. Just ran `npm run publish` again and now my dist directory looks like this:

image

It may be a local problem for me - maybe some npm caching - but you should probably try and see what happens when you do the same (add new stylesheets and partials).

File: scripts/forward-scss-files.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+function createForwardRule(filePath, packageAlias, sharedRootDir) {
+  return pipe(
+    (fileName) => (fileName[0] === '_' ? fileName.slice(1) : fileName),
+    (fileName) => (fileName === 'index.scss' ? '' : fileName),
+    (fileName) => path.basename(fileName, '.scss'),
+    (fileName) => `~${path.dirname(filePath).replace(sharedRootDir, packageAlias)}/${fileName}`,
+    (url) => (url[url.length - 1] === '/' ? url.slice(0, -1) : url),
+    (url) => `@forward "${url}";`
+  )(path.basename(filePath));
+}
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply scripts/forward-scss-files.js:22]: I will go with the second one 😄

File: scripts/forward-scss-files.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  return [...filesWithExtInDirectory, ...filesWithExtInSubdirectories.flat()];
+}
+
+function createForwardScssFile(
+  sourceFilePath,
+  sourceRootDir,
+  targetRootDir,
+  packageAlias,
+  sharedRootDir
+) {
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply scripts/forward-scss-files.js:47]: Good idea - i will have it use an object literal.

File: scripts/publish.js (hunk 4 above)
Comment (MadsBuchmann) [reply scripts/publish.js:124]: I'm not particularly a fan of passing in arguments to the function as it isn't done for any of the other functions in the ".then" chain at the bottom `publish..js`. 

How about renaming the step to `forwardCoreScssFiles` but having it take no arguments. Instead we will just have a call to `forwardScssFiles` for each library we want the files forwarded to. 

The `forwardCoreScssFiles` can then look something like: 
```
function forwardCoreScssFilesToDesignsystem() {
  console.log('Forwarding core SCSS files ...');
  const locations = [{from: 'sourcePath', to: 'targetPath'}, {from: 'sourcePath2', to: 'targetPath2'}]
  return new Promise((resolve) => {
     locations.forEach(({sourcePath, targetPath}) => {...}
   }
...
``

Comment (MadsBuchmann) [reply scripts/publish.js:124]: Ended up doing the following: 
```javascript
function forwardCoreScssFiles() {
  console.log('Forwarding core SCSS files...');

  return new Promise((resolve) => {
    pathsToForwardCoreScssFilesTo.forEach((targetPath) => {
      forwardScssFiles(`${coreLibDir}/scss`, targetPath, packageAlias, libsRootDir);
      resolve();
    });
  });
}
```

`pathsToForwardCoreScssFilesTo` can be found at the top of the file - closer to the variables it references.

File: scripts/forward-scss-files.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply scripts/forward-scss-files.js:22]: I've refactored the function to make it easier readable

**Review by jkaltoft — APPROVED** [review]

**Pull request merged**